tests/cases/unittests/session.ts

```diff
@@ -0,0 +1,459 @@
+					seq: 0,
+					request_seq: 0,
+					type: 'response',
+					command,
+					body: body,
+					success: true
+				});
+			});
+		});
+	});
+
+	describe('how Session is extendable via subclassing', () => {
+		let TestSession = class extends Session {
+			lastSent: protocol.Message;
+			customHandler: string = 'testhandler';
+			constructor(){
+				super(mockHost, Buffer.byteLength, process.hrtime, mockLogger);
+				this.addProtocolHandler(this.customHandler, () => {
+					return {response: undefined, responseRequired: true};
+				});
+			}
+			send(msg: protocol.Message) {
+				this.lastSent = msg;
+			}
+		};
+
+		it('can override methods such as send', () => {
+			let session = new TestSession(),
+			body = {
+				block: {
+					key: 'value'
+				}
+			},
+			command = 'test';
+
+			session.output(body, command);
+
+			expect(session.lastSent).to.deep.equal({
+				seq: 0,
+				request_seq: 0,
+				type: 'response',
+				command,
+				body: body,
+				success: true
+			});
+		});
+		it('can add and respond to new protocol handlers', () => {
+			let session = new TestSession();
+
+			expect(session.executeCommand({
+				seq: 0,
+				type: 'command',
+				command: session.customHandler
+			})).to.deep.equal({
+				response: undefined,
+				responseRequired: true
+			});
+		});
+		it('has access to the project service', () => {
+			let ServiceSession = class extends TestSession {
+				constructor() {
+					super();
+					assert(this.projectService);
+					expect(this.projectService).to.be.instanceOf(ProjectService);
+				}
+			};
+			new ServiceSession();
+		});
+	});
+
+	describe('an example of using the Session API to create an in-process server', () => {
+		let inProcHost: ServerHost = {
+			args: [],
+			newLine: '\n',
+			useCaseSensitiveFileNames: true,
+			write: (s) => lastWrittenToHost = s,
+			readFile: () => void 0,
+			writeFile: () => void 0,
+			resolvePath: () => void 0,
+			fileExists: () => false,
+			directoryExists: () => false,
+			createDirectory: () => void 0,
+			getExecutingFilePath: () => void 0,
+			getCurrentDirectory: () => void 0,
+			readDirectory: () => void 0,
+			exit: () => void 0
+		},
+		InProcSession = class extends Session {
+			private queue: protocol.Request[] = [];
+			constructor(private client: {handle: (msg: protocol.Message) => void}) {
+				super(inProcHost, Buffer.byteLength, process.hrtime, mockLogger);
+				this.addProtocolHandler('echo', (req: protocol.Request) => ({
+					response: req.arguments,
+					responseRequired: true
+				}));
+			}
+			
+			send(msg: protocol.Message) {
+				this.client.handle(msg);
+			}
+
+			enqueue(msg: protocol.Request) {
+				this.queue = [msg].concat(this.queue);
+			}
+			
+			handleRequest(msg: protocol.Request) {
+				let response: protocol.Response;
+				try {
+					response = this.executeCommand(msg).response;
+				} catch (e) {
+					this.output(undefined, msg.command, msg.seq, e.toString());
+					return;
+				}
+				if (response) {

```

Comment:
Yes. `response` can be `undefined` while `responseRequired` is true. (Indicating that it should pretty much just ack the message.) This is done in the server.